File: qa/src/http.mjs
/**
 * Read-only HTTP layer.
 *
 * Every request made by this suite goes through here, which guarantees:
 *  - GET/HEAD only (never POST/PUT/DELETE against the live site)
 *  - full redirect-chain capture (manual redirect following)
 *  - a polite delay + bounded concurrency
 *  - one fetch per normalised URL (response cache)
 */

const METHOD_ALLOWLIST = new Set(['GET', 'HEAD']);

export class Http {
  constructor(config) {
    this.cfg = config;
    this.cache = new Map(); // normalisedUrl -> result
    this.robots = null;
    this.stats = { requests: 0, bytes: 0, errors: 0 };
    this._lastRequestAt = 0;
  }

  /** Strip tracking params, drop fragments, normalise trailing slash + host case. */
  normalise(input, base = this.cfg.site.baseUrl) {
    let u;
    try {
      u = new URL(input, base);
    } catch {
      return null;
    }
    if (!/^https?:$/.test(u.protocol)) return null;
    u.hash = '';
    u.hostname = u.hostname.toLowerCase();
    for (const p of this.cfg.crawl.stripParams) u.searchParams.delete(p);
    // Sort remaining params so ?a=1&b=2 and ?b=2&a=1 are one URL.
    u.searchParams.sort();
    if (u.pathname !== '/' && u.pathname.endsWith('/') && !this.cfg.site.canonicalTrailingSlash) {
      u.pathname = u.pathname.replace(/\/+$/, '');
    }
    return u.toString();
  }

  isInternal(url) {
    try {
      const host = new URL(url).hostname.toLowerCase();
      return this.cfg.site.internalHosts.includes(host);
    } catch {
      return false;
    }
  }

  isIgnored(url) {
    // Fragments are already stripped by normalise(); test the path only.
    let path;
    try {
      const u = new URL(url);
      path = u.pathname + u.search;
    } catch {
      return true;
    }
    return this.cfg.crawl.ignore.some((re) => re.test(path));
  }

  async _throttle() {
    const wait = this.cfg.crawl.delayMs - (Date.now() - this._lastRequestAt);
    if (wait > 0) await new Promise((r) => setTimeout(r, wait));
    this._lastRequestAt = Date.now();
  }

  /**
   * Fetch a URL, following redirects by hand so the whole chain is recorded.
   * Returns { url, finalUrl, status, ok, chain, headers, body, contentType, timingMs, error }.
   */
  async fetchWithChain(rawUrl, { method = 'GET', maxHops = 10, wantBody = true } = {}) {
    if (!METHOD_ALLOWLIST.has(method)) {
      throw new Error(`Refusing non-read-only method: ${method}`);
    }
    const url = this.normalise(rawUrl);
    if (!url) {
      return { url: rawUrl, finalUrl: rawUrl, status: 0, ok: false, chain: [], headers: {}, body: '', error: 'invalid-url' };
    }

    const cacheKey = `${method}:${url}`;
    if (this.cache.has(cacheKey)) return this.cache.get(cacheKey);

    const chain = [];
    let current = url;
    let result = null;
    const started = Date.now();

    for (let hop = 0; hop <= maxHops; hop++) {
      const res = await this._raw(current, method);
      if (res.error) {
        result = {
          url, finalUrl: current, status: 0, ok: false, chain,
          headers: {}, body: '', contentType: '', timingMs: Date.now() - started,
          error: res.error,
        };
        break;
      }
      const status = res.status;
      const location = res.headers.get('location');
      if (status >= 300 && status < 400 && location) {
        const next = this.normalise(location, current);
        chain.push({ from: current, to: next, status });
        if (!next || next === current) {
          result = { url, finalUrl: current, status, ok: false, chain, headers: headersToObject(res.headers), body: '', contentType: '', timingMs: Date.now() - started, error: 'redirect-loop' };
          break;
        }
        current = next;
        continue;
      }

      let body = '';
      const contentType = res.headers.get('content-type') || '';
      if (wantBody && method === 'GET' && /text\/|json|xml|javascript|\+xml/i.test(contentType)) {
        try {
          body = await res.text();
          this.stats.bytes += body.length;
        } catch {
          body = '';
        }
      } else if (res.body) {
        try { await res.arrayBuffer(); } catch { /* drain */ }
      }

      result = {
        url,
        finalUrl: current,
        status,
        ok: status >= 200 && status < 300,
        chain,
        headers: headersToObject(res.headers),
        body,
        contentType,
        timingMs: Date.now() - started,
        error: null,
      };
      break;
    }

    if (!result) {
      result = { url, finalUrl: current, status: 0, ok: false, chain, headers: {}, body: '', contentType: '', timingMs: Date.now() - started, error: 'too-many-redirects' };
    }
    if (!result.ok) this.stats.errors++;
    this.cache.set(cacheKey, result);
    return result;
  }

  async _raw(url, method) {
    const { retries, timeoutMs, userAgent } = this.cfg.crawl;
    let lastError = null;
    for (let attempt = 0; attempt <= retries; attempt++) {
      await this._throttle();
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), timeoutMs);
      try {
        this.stats.requests++;
        const res = await fetch(url, {
          method,
          redirect: 'manual',
          signal: controller.signal,
          headers: {
            'user-agent': userAgent,
            accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
            'accept-language': 'en-AU,en;q=0.9',
          },
        });
        clearTimeout(timer);
        return { status: res.status, headers: res.headers, body: res.body, text: () => res.text(), arrayBuffer: () => res.arrayBuffer() };
      } catch (err) {
        clearTimeout(timer);
        lastError = err?.name === 'AbortError' ? `timeout after ${timeoutMs}ms` : (err?.cause?.code || err?.message || String(err));
        if (attempt < retries) {
          await new Promise((r) => setTimeout(r, 500 * 2 ** attempt));
        }
      }
    }
    return { error: lastError };
  }

  /** Fetch and parse robots.txt once. */
  async loadRobots() {
    if (this.robots) return this.robots;
    const res = await this.fetchWithChain(new URL('/robots.txt', this.cfg.site.baseUrl).toString());
    this.robots = parseRobots(res.ok ? res.body : '', res);
    return this.robots;
  }

  /** True when robots.txt permits our user-agent to fetch this path. */
  allowedByRobots(url) {
    if (!this.cfg.crawl.respectRobots || !this.robots) return true;
    if (!this.isInternal(url)) return true;
    let path;
    try {
      const u = new URL(url);
      path = u.pathname + u.search;
    } catch {
      return true;
    }
    return this.robots.isAllowed(path);
  }
}

function headersToObject(headers) {
  const out = {};
  for (const [k, v] of headers.entries()) out[k.toLowerCase()] = v;
  return out;
}

/**
 * Minimal robots.txt parser: enough for allow/disallow on the `*` group plus
 * Sitemap discovery. Longest-match wins, Allow beats Disallow on a tie.
 */
export function parseRobots(text, response = null) {
  const groups = [];
  const sitemaps = [];
  let current = null;

  for (const rawLine of String(text).split(/\r?\n/)) {
    const line = rawLine.replace(/#.*$/, '').trim();
    if (!line) continue;
    const m = line.match(/^([a-zA-Z-]+)\s*:\s*(.*)$/);
    if (!m) continue;
    const field = m[1].toLowerCase();
    const value = m[2].trim();

    if (field === 'user-agent') {
      if (!current || current.rules.length) {
        current = { agents: [], rules: [] };
        groups.push(current);
      }
      current.agents.push(value.toLowerCase());
    } else if (field === 'disallow' || field === 'allow') {
      if (!current) {
        current = { agents: ['*'], rules: [] };
        groups.push(current);
      }
      current.rules.push({ type: field, path: value });
    } else if (field === 'sitemap') {
      sitemaps.push(value);
    }
  }

  const group = groups.find((g) => g.agents.includes('*')) || { rules: [] };

  function matchLength(pattern, path) {
    if (pattern === '') return -1;
    // Support the * wildcard and $ end-anchor used by robots.txt.
    const hasWildcard = pattern.includes('*') || pattern.endsWith('$');
    if (!hasWildcard) return path.startsWith(pattern) ? pattern.length : -1;
    const re = new RegExp(
      '^' + pattern.replace(/[.+?^${}()|[\]\\]/g, '\\$&').replace(/\*/g, '.*').replace(/\\\$$/, '$'),
    );
    return re.test(path) ? pattern.length : -1;
  }

  return {
    raw: String(text),
    ok: response ? response.ok : true,
    status: response ? response.status : 0,
    sitemaps,
    groups,
    /** Whole-site block: `Disallow: /` with no offsetting Allow. */
    blocksEverything() {
      const disallowRoot = group.rules.some((r) => r.type === 'disallow' && r.path === '/');
      const allowRoot = group.rules.some((r) => r.type === 'allow' && r.path === '/');
      return disallowRoot && !allowRoot;
    },
    isAllowed(path) {
      let best = { type: 'allow', len: -1 };
      for (const rule of group.rules) {
        const len = matchLength(rule.path, path);
        if (len > best.len || (len === best.len && rule.type === 'allow')) {
          best = { type: rule.type, len };
        }
      }
      return best.len < 0 || best.type === 'allow';
    },
  };
}

File: qa/emjay_qa/http.py
"""Read-only HTTP layer.

Two guarantees this module enforces so the suite can never alter the live site:

1. Only GET and HEAD are ever sent. Any other method raises.
2. Redirects are followed manually so every hop is recorded, and the hop limit
   is hard-capped.
"""
from __future__ import annotations

import time
import urllib.robotparser
from typing import Dict, Optional
from urllib.parse import urlparse

import requests

from .config import Config
from .models import FetchResult, RobotsInfo

SAFE_METHODS = {"GET", "HEAD"}
MAX_HOPS = 10
MAX_BODY_BYTES = 8 * 1024 * 1024


class UnsafeMethodError(RuntimeError):
    pass


class Fetcher:
    def __init__(self, config: Config):
        self.config = config
        self.session = requests.Session()
        self.session.headers.update(
            {
                "User-Agent": config.get("crawl.user_agent", "EmjayQA/1.0"),
                "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                "Accept-Language": "en-AU,en;q=0.9",
                "Accept-Encoding": "gzip, deflate",
            }
        )
        self.timeout = float(config.get("crawl.timeout_seconds", 25))
        self.delay = float(config.get("crawl.delay_seconds", 0.7))
        self.verify = bool(config.get("crawl.verify_tls", True))
        self._cache: Dict[str, FetchResult] = {}
        self._last_hit: Dict[str, float] = {}
        self._robots: Dict[str, urllib.robotparser.RobotFileParser] = {}
        self.request_count = 0
        self.methods_used: set = set()

    # ------------------------------------------------------------------ core
    def _throttle(self, url: str) -> None:
        host = urlparse(url).netloc
        last = self._last_hit.get(host)
        if last is not None:
            wait = self.delay - (time.time() - last)
            if wait > 0:
                time.sleep(wait)
        self._last_hit[host] = time.time()

    def fetch(
        self,
        url: str,
        method: str = "GET",
        want_body: bool = True,
        use_cache: bool = True,
    ) -> FetchResult:
        method = method.upper()
        if method not in SAFE_METHODS:
            raise UnsafeMethodError(f"{method} is not permitted: this suite is read-only")
        self.methods_used.add(method)

        cache_key = f"{method}:{url}:{int(want_body)}"
        if use_cache and cache_key in self._cache:
            return self._cache[cache_key]

        result = FetchResult(url=url)
        current = url
        started = time.time()
        try:
            for _ in range(MAX_HOPS):
                self._throttle(current)
                self.request_count += 1
                response = self.session.request(
                    method,
                    current,
                    allow_redirects=False,
                    timeout=self.timeout,
                    verify=self.verify,
                    stream=True,
                )
                result.chain.append((current, response.status_code))
                if response.is_redirect or response.status_code in (301, 302, 303, 307, 308):
                    location = response.headers.get("Location")
                    response.close()
                    if not location:
                        result.error = f"{response.status_code} with no Location header"
                        break
                    current = requests.compat.urljoin(current, location)
                    continue

                result.status = response.status_code
                result.final_url = current
                result.headers = {k.lower(): v for k, v in response.headers.items()}
                result.content_type = result.headers.get("content-type", "")
                body = b""
                if want_body and method == "GET":
                    for chunk in response.iter_content(65536):
                        body += chunk
                        if len(body) > MAX_BODY_BYTES:
                            result.error = "body truncated at 8MB"
                            break
                    result.bytes = len(body)
                    lowered = (result.content_type or "").lower()
                    textual = result.is_html or any(
                        kind in lowered for kind in ("xml", "json", "css", "javascript", "text/")
                    )
                    if textual:
                        encoding = response.encoding or "utf-8"
                        try:
                            result.text = body.decode(encoding, errors="replace")
                        except LookupError:
                            result.text = body.decode("utf-8", errors="replace")
                else:
                    result.bytes = int(result.headers.get("content-length") or 0)
                response.close()
                break
            else:
                result.error = f"redirect loop: more than {MAX_HOPS} hops"
        except requests.RequestException as exc:
            result.error = f"{type(exc).__name__}: {exc}"
        except Exception as exc:  # pragma: no cover - defensive
            result.error = f"{type(exc).__name__}: {exc}"

        if result.status is None and result.chain and not result.error:
            result.status = result.chain[-1][1]
            result.final_url = result.chain[-1][0]
        result.final_url = result.final_url or current
        result.elapsed_ms = (time.time() - started) * 1000.0
        if use_cache:
            self._cache[cache_key] = result
        return result

    # --------------------------------------------------------------- robots
    def robots(self, base_url: str) -> RobotsInfo:
        robots_url = requests.compat.urljoin(base_url + "/", "/robots.txt")
        result = self.fetch(robots_url)
        info = RobotsInfo(url=robots_url, status=result.status, error=result.error)
        text = result.text or ""
        info.text = text
        for line in text.splitlines():
            stripped = line.strip()
            if stripped.lower().startswith("sitemap:"):
                info.sitemaps.append(stripped.split(":", 1)[1].strip())
        parser = urllib.robotparser.RobotFileParser()
        parser.parse(text.splitlines())
        self._robots[urlparse(base_url).netloc] = parser
        agent = self.config.get("crawl.user_agent", "EmjayQA/1.0")
        info.disallow_all = bool(text.strip()) and not parser.can_fetch(agent, base_url + "/")
        return info

    def allowed_by_robots(self, url: str) -> bool:
        if not self.config.get("crawl.obey_robots", True):
            return True
        parser = self._robots.get(urlparse(url).netloc)
        if parser is None:
            return True
        return parser.can_fetch(self.config.get("crawl.user_agent", "EmjayQA/1.0"), url)

    def cached(self, url: str) -> Optional[FetchResult]:
        for key, value in self._cache.items():
            if key.split(":", 1)[1].rsplit(":", 1)[0] == url:
                return value
        return None
